xnet/util.go

```diff
@@ -0,0 +1,52 @@
+package xnet
+
+import (
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/Sirupsen/logrus"
+)
+
+// ParseProtoAddr parses an address as a protcol and address pair.
+// Try its best to guess and fallback to return Unix's socket.
+func ParseProtoAddr(s string) (string, string) {
+	logrus.Debugf("parse proto addr: %s", s)
+	parts := strings.SplitN(s, "://", 2)
+	if len(parts) == 1 {
+		_, _, err := net.SplitHostPort(parts[0])
+		if err == nil {
+			return "tcp", parts[0]
+		}
+
+		return "unix", parts[0]
+	}
+	return parts[0], parts[1]
+}
+
+// ResolveLocalAddr resolves for the local address.
+func ResolveLocalAddr(remoteAddr string) (string, error) {
+	if listenHost, listenPort, err := net.SplitHostPort(remoteAddr); err == nil {
+		ip := net.ParseIP(listenHost)
+		if ip != nil && ip.IsUnspecified() {
+			// Find our local IP address associated with the default route.
+			// This may not be the appropriate address to use for internal
+			// cluster communications, but it seems like the best default.
+			// The admin can override this address if necessary.
+			conn, err := net.Dial("udp", "8.8.8.8:53")
```

Comment:
I'm not sure if we should cement this is the methodology for resolving a node's external address. This won't work correctly in a number of situations and will be somewhat challenging to debug.